{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Train"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# screen -dmS 4dgs_Case_I -L -Logfile 4dgs_Case_I.log sh -c 'CUDA_VISIBLE_DEVICES=0 python 4DGaussians/train.py -s data_ns/Case_I --port 6014 --expname \"4dgs/Case_I\" --configs 4DGaussians/arguments/hypernerf/default.py'\n",
    "# nvidia-smi dmon -s umc -i 0 -d 1 -f 4dgs_Case_I_gpu_util.log &\n",
    "# 273 4163\n",
    "# screen -dmS 4dgs_Case_II -L -Logfile 4dgs_Case_II.log sh -c 'CUDA_VISIBLE_DEVICES=1 python 4DGaussians/train.py -s data_ns/Case_II --port 6015 --expname \"4dgs/Case_II\" --configs 4DGaussians/arguments/hypernerf/default.py'\n",
    "# nvidia-smi dmon -s umc -i 1 -d 1 -f 4dgs_Case_II_gpu_util.log &\n",
    "# 12 4242\n",
    "# screen -dmS 4dgs_Case_III -L -Logfile 4dgs_Case_III.log sh -c 'CUDA_VISIBLE_DEVICES=0 python 4DGaussians/train.py -s data_ns/Case_III --port 6016 --expname \"4dgs/Case_III\" --configs 4DGaussians/arguments/hypernerf/default.py'\n",
    "# nvidia-smi dmon -s umc -i 0 -d 1 -f 4dgs_Case_III_gpu_util.log &\n",
    "\n",
    "# screen -dmS 4dgs_Case_IV -L -Logfile 4dgs_Case_IV.log sh -c 'CUDA_VISIBLE_DEVICES=1 python 4DGaussians/train.py -s data_ns/Case_IV --port 6017 --expname \"4dgs/Case_IV\" --configs 4DGaussians/arguments/hypernerf/default.py'\n",
    "# nvidia-smi dmon -s umc -i 1 -d 1 -f 4dgs_Case_IV_gpu_util.log &\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Render"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Looking for config file in models/4dgs/Case_I/cfg_args\n",
      "Config file found: models/4dgs/Case_I/cfg_args\n",
      "Rendering  models/4dgs/Case_I\n",
      "feature_dim: 48 [25/02 20:48:22]\n",
      "Loading trained model at iteration 14000 [25/02 20:48:22]\n",
      "Found transforms_train.json file, assuming Blender data set! [25/02 20:48:22]\n",
      "Reading Training Transforms [25/02 20:48:22]\n",
      "Reading Test Transforms [25/02 20:48:22]\n",
      "Generating Video Transforms [25/02 20:48:22]\n",
      "hello!!!! [25/02 20:48:22]\n",
      "Generating random point cloud (2000)... [25/02 20:48:22]\n",
      "Loading Training Cameras [25/02 20:48:22]\n",
      "Loading Test Cameras [25/02 20:48:22]\n",
      "Loading Video Cameras [25/02 20:48:22]\n",
      "Deformation Net Set aabb [1.29982098 1.29990645 1.29988719] [-1.29980838 -1.29981163 -1.29872349] [25/02 20:48:22]\n",
      "Voxel Plane: set aabb= Parameter containing:\n",
      "tensor([[ 1.2998,  1.2999,  1.2999],\n",
      "        [-1.2998, -1.2998, -1.2987]]) [25/02 20:48:22]\n",
      "loading model from existsmodels/4dgs/Case_I/point_cloud/iteration_14000 [25/02 20:48:23]\n",
      "point nums: 149300 [25/02 20:48:23]\n",
      "Rendering progress: 100%|█████████████████████████| 2/2 [00:00<00:00, 19.19it/s]\n",
      "Average FPS: 46.0060152289275 [25/02 20:48:25]\n",
      "Traceback (most recent call last):\n",
      "  File \"4DGaussians/render.py\", line 136, in <module>\n",
      "    render_sets(model.extract(args), hyperparam.extract(args), args.iteration, pipeline.extract(args), args.skip_train, args.skip_test, args.skip_video)\n",
      "  File \"4DGaussians/render.py\", line 111, in render_sets\n",
      "    render_set(dataset.model_path, \"test\", scene.loaded_iter, scene.getTestCameras(), gaussians, pipeline, background,cam_type)\n",
      "  File \"4DGaussians/render.py\", line 98, in render_set\n",
      "    imageio.mimwrite(os.path.join(model_path, name, \"ours_{}\".format(iteration), 'video_rgb.mp4'), render_images, fps=30)\n",
      "  File \"/home/rwhite/mambaforge/envs/Gaussians4D/lib/python3.7/site-packages/imageio/v2.py\", line 494, in mimwrite\n",
      "    with imopen(uri, \"wI\", **imopen_args) as file:\n",
      "  File \"/home/rwhite/mambaforge/envs/Gaussians4D/lib/python3.7/site-packages/imageio/core/imopen.py\", line 281, in imopen\n",
      "    raise err_type(err_msg)\n",
      "ValueError: Could not find a backend to open `models/4dgs/Case_I/test/ours_14000/video_rgb.mp4`` with iomode `wI`.\n",
      "Based on the extension, the following plugins might add capable backends:\n",
      "  FFMPEG:  pip install imageio[ffmpeg]\n",
      "  pyav:  pip install imageio[pyav]\n",
      "Looking for config file in models/4dgs/Case_II/cfg_args\n",
      "Config file found: models/4dgs/Case_II/cfg_args\n",
      "Rendering  models/4dgs/Case_II\n",
      "feature_dim: 48 [25/02 20:48:30]\n",
      "Loading trained model at iteration 14000 [25/02 20:48:30]\n",
      "Found transforms_train.json file, assuming Blender data set! [25/02 20:48:30]\n",
      "Reading Training Transforms [25/02 20:48:30]\n",
      "Reading Test Transforms [25/02 20:48:30]\n",
      "Generating Video Transforms [25/02 20:48:30]\n",
      "hello!!!! [25/02 20:48:30]\n",
      "Generating random point cloud (2000)... [25/02 20:48:30]\n",
      "Loading Training Cameras [25/02 20:48:30]\n",
      "Loading Test Cameras [25/02 20:48:30]\n",
      "Loading Video Cameras [25/02 20:48:30]\n",
      "Deformation Net Set aabb [1.29982098 1.29990645 1.29988719] [-1.29980838 -1.29981163 -1.29872349] [25/02 20:48:30]\n",
      "Voxel Plane: set aabb= Parameter containing:\n",
      "tensor([[ 1.2998,  1.2999,  1.2999],\n",
      "        [-1.2998, -1.2998, -1.2987]]) [25/02 20:48:30]\n",
      "loading model from existsmodels/4dgs/Case_II/point_cloud/iteration_14000 [25/02 20:48:31]\n",
      "point nums: 157956 [25/02 20:48:31]\n",
      "Rendering progress: 100%|█████████████████████████| 3/3 [00:00<00:00, 23.61it/s]\n",
      "Average FPS: 62.52642720313653 [25/02 20:48:34]\n",
      "Traceback (most recent call last):\n",
      "  File \"4DGaussians/render.py\", line 136, in <module>\n",
      "    render_sets(model.extract(args), hyperparam.extract(args), args.iteration, pipeline.extract(args), args.skip_train, args.skip_test, args.skip_video)\n",
      "  File \"4DGaussians/render.py\", line 111, in render_sets\n",
      "    render_set(dataset.model_path, \"test\", scene.loaded_iter, scene.getTestCameras(), gaussians, pipeline, background,cam_type)\n",
      "  File \"4DGaussians/render.py\", line 98, in render_set\n",
      "    imageio.mimwrite(os.path.join(model_path, name, \"ours_{}\".format(iteration), 'video_rgb.mp4'), render_images, fps=30)\n",
      "  File \"/home/rwhite/mambaforge/envs/Gaussians4D/lib/python3.7/site-packages/imageio/v2.py\", line 494, in mimwrite\n",
      "    with imopen(uri, \"wI\", **imopen_args) as file:\n",
      "  File \"/home/rwhite/mambaforge/envs/Gaussians4D/lib/python3.7/site-packages/imageio/core/imopen.py\", line 281, in imopen\n",
      "    raise err_type(err_msg)\n",
      "ValueError: Could not find a backend to open `models/4dgs/Case_II/test/ours_14000/video_rgb.mp4`` with iomode `wI`.\n",
      "Based on the extension, the following plugins might add capable backends:\n",
      "  FFMPEG:  pip install imageio[ffmpeg]\n",
      "  pyav:  pip install imageio[pyav]\n",
      "Looking for config file in models/4dgs/Case_III/cfg_args\n",
      "Config file found: models/4dgs/Case_III/cfg_args\n",
      "Rendering  models/4dgs/Case_III\n",
      "feature_dim: 48 [25/02 20:48:39]\n",
      "Loading trained model at iteration 14000 [25/02 20:48:39]\n",
      "Found transforms_train.json file, assuming Blender data set! [25/02 20:48:39]\n",
      "Reading Training Transforms [25/02 20:48:39]\n",
      "Reading Test Transforms [25/02 20:48:42]\n",
      "Generating Video Transforms [25/02 20:48:42]\n",
      "hello!!!! [25/02 20:48:42]\n",
      "Generating random point cloud (2000)... [25/02 20:48:42]\n",
      "Loading Training Cameras [25/02 20:48:42]\n",
      "Loading Test Cameras [25/02 20:48:42]\n",
      "Loading Video Cameras [25/02 20:48:42]\n",
      "Deformation Net Set aabb [1.29982098 1.29990645 1.29988719] [-1.29980838 -1.29981163 -1.29872349] [25/02 20:48:42]\n",
      "Voxel Plane: set aabb= Parameter containing:\n",
      "tensor([[ 1.2998,  1.2999,  1.2999],\n",
      "        [-1.2998, -1.2998, -1.2987]]) [25/02 20:48:42]\n",
      "loading model from existsmodels/4dgs/Case_III/point_cloud/iteration_14000 [25/02 20:48:43]\n",
      "point nums: 3616 [25/02 20:48:43]\n",
      "Rendering progress: 100%|███████████████████████| 10/10 [00:00<00:00, 53.45it/s]\n",
      "Average FPS: 159.78179213872656 [25/02 20:48:49]\n",
      "Traceback (most recent call last):\n",
      "  File \"4DGaussians/render.py\", line 136, in <module>\n",
      "    render_sets(model.extract(args), hyperparam.extract(args), args.iteration, pipeline.extract(args), args.skip_train, args.skip_test, args.skip_video)\n",
      "  File \"4DGaussians/render.py\", line 111, in render_sets\n",
      "    render_set(dataset.model_path, \"test\", scene.loaded_iter, scene.getTestCameras(), gaussians, pipeline, background,cam_type)\n",
      "  File \"4DGaussians/render.py\", line 98, in render_set\n",
      "    imageio.mimwrite(os.path.join(model_path, name, \"ours_{}\".format(iteration), 'video_rgb.mp4'), render_images, fps=30)\n",
      "  File \"/home/rwhite/mambaforge/envs/Gaussians4D/lib/python3.7/site-packages/imageio/v2.py\", line 494, in mimwrite\n",
      "    with imopen(uri, \"wI\", **imopen_args) as file:\n",
      "  File \"/home/rwhite/mambaforge/envs/Gaussians4D/lib/python3.7/site-packages/imageio/core/imopen.py\", line 281, in imopen\n",
      "    raise err_type(err_msg)\n",
      "ValueError: Could not find a backend to open `models/4dgs/Case_III/test/ours_14000/video_rgb.mp4`` with iomode `wI`.\n",
      "Based on the extension, the following plugins might add capable backends:\n",
      "  FFMPEG:  pip install imageio[ffmpeg]\n",
      "  pyav:  pip install imageio[pyav]\n",
      "Looking for config file in models/4dgs/Case_IV/cfg_args\n",
      "Config file found: models/4dgs/Case_IV/cfg_args\n",
      "Rendering  models/4dgs/Case_IV\n",
      "feature_dim: 48 [25/02 20:48:53]\n",
      "Loading trained model at iteration 14000 [25/02 20:48:53]\n",
      "Found transforms_train.json file, assuming Blender data set! [25/02 20:48:53]\n",
      "Reading Training Transforms [25/02 20:48:53]\n",
      "Reading Test Transforms [25/02 20:48:53]\n",
      "Generating Video Transforms [25/02 20:48:53]\n",
      "hello!!!! [25/02 20:48:53]\n",
      "Generating random point cloud (2000)... [25/02 20:48:53]\n",
      "Loading Training Cameras [25/02 20:48:53]\n",
      "Loading Test Cameras [25/02 20:48:53]\n",
      "Loading Video Cameras [25/02 20:48:53]\n",
      "Deformation Net Set aabb [1.29982098 1.29990645 1.29988719] [-1.29980838 -1.29981163 -1.29872349] [25/02 20:48:53]\n",
      "Voxel Plane: set aabb= Parameter containing:\n",
      "tensor([[ 1.2998,  1.2999,  1.2999],\n",
      "        [-1.2998, -1.2998, -1.2987]]) [25/02 20:48:53]\n",
      "loading model from existsmodels/4dgs/Case_IV/point_cloud/iteration_14000 [25/02 20:48:54]\n",
      "point nums: 282795 [25/02 20:48:54]\n",
      "Rendering progress: 100%|█████████████████████████| 2/2 [00:00<00:00, 15.89it/s]\n",
      "Average FPS: 25.521646799681154 [25/02 20:48:58]\n",
      "Traceback (most recent call last):\n",
      "  File \"4DGaussians/render.py\", line 136, in <module>\n",
      "    render_sets(model.extract(args), hyperparam.extract(args), args.iteration, pipeline.extract(args), args.skip_train, args.skip_test, args.skip_video)\n",
      "  File \"4DGaussians/render.py\", line 111, in render_sets\n",
      "    render_set(dataset.model_path, \"test\", scene.loaded_iter, scene.getTestCameras(), gaussians, pipeline, background,cam_type)\n",
      "  File \"4DGaussians/render.py\", line 98, in render_set\n",
      "    imageio.mimwrite(os.path.join(model_path, name, \"ours_{}\".format(iteration), 'video_rgb.mp4'), render_images, fps=30)\n",
      "  File \"/home/rwhite/mambaforge/envs/Gaussians4D/lib/python3.7/site-packages/imageio/v2.py\", line 494, in mimwrite\n",
      "    with imopen(uri, \"wI\", **imopen_args) as file:\n",
      "  File \"/home/rwhite/mambaforge/envs/Gaussians4D/lib/python3.7/site-packages/imageio/core/imopen.py\", line 281, in imopen\n",
      "    raise err_type(err_msg)\n",
      "ValueError: Could not find a backend to open `models/4dgs/Case_IV/test/ours_14000/video_rgb.mp4`` with iomode `wI`.\n",
      "Based on the extension, the following plugins might add capable backends:\n",
      "  FFMPEG:  pip install imageio[ffmpeg]\n",
      "  pyav:  pip install imageio[pyav]\n"
     ]
    }
   ],
   "source": [
    "!python 4DGaussians/render.py --model_path \"models/4dgs/Case_I\"  --skip_train --configs 4DGaussians/arguments/hypernerf/default.py --skip_video\n",
    "!python 4DGaussians/render.py --model_path \"models/4dgs/Case_II\"  --skip_train --configs 4DGaussians/arguments/hypernerf/default.py --skip_video\n",
    "!python 4DGaussians/render.py --model_path \"models/4dgs/Case_III\"  --skip_train --configs 4DGaussians/arguments/hypernerf/default.py --skip_video\n",
    "!python 4DGaussians/render.py --model_path \"models/4dgs/Case_IV\"  --skip_train --configs 4DGaussians/arguments/hypernerf/default.py --skip_video"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Gaussians4D",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
